### PR TITLE
DS-3851 Endpoint to interact with the workflow

### PR DESCRIPTION
**Pull request by abollini**
This PR is a placeholder to share early the "second part" of https://github.com/DSpace/DSpace/pull/2102

It enable the configurable workflow and implement the workflowItem endpoint providing the ability to claim, unclaim and approve tasks trasforming an inprogress submission in a "final item".

The enhancement to discovery to index inprogress submission enabling the new mydspace will be introduced by a separate PR

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/content/dao/ItemDAO.java [hunk 1]
```diff
@@ -131,4 +133,12 @@ public int countItems(Context context, List<Collection> collections, boolean inc
      */
     int countItems(Context context, boolean includeArchived, boolean includeWithdrawn) throws SQLException;
 
+    public List<Bundle> findBundlesByName(Context context, Item item, String name) throws SQLException;
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/dao/ItemDAO.java]: Please add JavaDocs for these 3 new public methods (in this same class)

File: dspace-api/src/main/java/org/dspace/content/factory/ContentServiceFactoryImpl.java [hunk 2]
```diff
@@ -35,7 +37,9 @@
  */
 public class ContentServiceFactoryImpl extends ContentServiceFactory {
 
-
+//    @Autowired(required = true)
+//    private List<BrowsableObjectService<? extends BrowsableDSpaceObject<? extends Serializable>, Serializable>>
+//        browsableDSpaceObjectServices;
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/factory/ContentServiceFactoryImpl.java]: These commented out lines likely should be removed

File: dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java [hunk 3]
```diff
@@ -249,10 +237,16 @@ public void addMetadata(Context context, T dso, String schema, String element, S
      * @throws SQLException if database error
      */
     public void addMetadata(Context context, T dso, MetadataField metadataField, String lang, List<String> values,
-                            List<String> authorities, List<Integer> confidences) throws SQLException;
+            List<String> authorities, List<Integer> confidences) throws SQLException;
+
+    public void addMetadata(Context context, T dso, MetadataField metadataField, String lang, List<String> values,
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java]: Add Javadocs

File: dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java [hunk 4]
```diff
@@ -249,10 +237,16 @@ public void addMetadata(Context context, T dso, String schema, String element, S
...
+            List<String> authorities, List<Integer> confidences, List<Integer> places) throws SQLException;
 
     public void addMetadata(Context context, T dso, MetadataField metadataField, String language, String value,
-                            String authority, int confidence) throws SQLException;
+            String authority, int confidence) throws SQLException;
+
+    public void addMetadata(Context context, T dso, MetadataField metadataField, String language, String value,
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java]: Add Javadocs

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java [hunk 5]
```diff
@@ -37,6 +37,9 @@
  * @author kevinvandevelde at atmire.com
  */
 public interface ItemService extends DSpaceObjectService<Item>, DSpaceObjectLegacySupportService<Item> {
+
+    public Item find(Context context, UUID id) throws SQLException;
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/ItemService.java]: Add javadocs for the 5 new methods in this Class

File: dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItem.java [hunk 6]
```diff
@@ -183,4 +187,33 @@ public boolean isPublishedBefore() {
...
+        } finally {
+            if (context != null && context.isValid()) {
+                context.abort();
+            }
+        }
+    }
+
+    @Override
+    public String getTypeText() {
+        return "workflowitem";
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItem.java]: Should this be in `Constants`?

File: dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java [hunk 7]
```diff
@@ -1041,4 +1058,62 @@ protected void notifyOfReject(Context c, XmlWorkflowItem wi, EPerson e,
     public String getMyDSpaceLink() {
         return ConfigurationManager.getProperty("dspace.url") + "/mydspace";
     }
+
+    public void createMessage(Context context, Item item, String description, String subject)
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java]: Add Javadocs please

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/ClaimedTask.java [hunk 8]
```diff
@@ -115,4 +123,29 @@ public void setWorkflowID(String workflowID) {
     public String getWorkflowID() {
         return workflowId;
     }
+
+    @Override
+    public String getTypeText() {
+        return "claimedtask";
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/ClaimedTask.java]: Should this be in `Constants`?

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/PoolTask.java [hunk 9]
```diff
@@ -130,4 +138,29 @@ public void setActionID(String actionID) {
     public String getActionID() {
         return this.actionId;
     }
+
+    @Override
+    public String getTypeText() {
+        return "pooltask";
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/PoolTask.java]: Should this be in `Constants`?

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java [hunk 10]
```diff
@@ -134,4 +139,49 @@ public void setPublishedBefore(boolean b) {
...
+
+    @Override
+    public int getState() {
+        // FIXME
+        return 0;
+    }
+
+    @Override
+    public String getTypeText() {
+        return "workflowitem";
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java]: Should this be in `Constants`?

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java [hunk 11]
```diff
@@ -134,4 +139,49 @@ public void setPublishedBefore(boolean b) {
...
+            if (context != null && context.isValid()) {
+                context.abort();
+            }
+        }
+    }
+
+    @Override
+    public int getState() {
+        // FIXME
+        return 0;
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java]: Please explain this "FIXME" further.  How should this be fixed/improved?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/BrowsableDSpaceObjectConverter.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ * the REST data model that can be indexed
+ *
+ * @param <M> the Class in the DSpace API data model
+ * @param <R> the Class in the DSpace REST data model
+ * @author Andrea Bollini (andrea.bollini at 4science.it)
+ */
+public abstract class BrowsableDSpaceObjectConverter<M extends BrowsableDSpaceObject,
+    R extends org.dspace.app.rest.model.RestAddressableModel> extends DSpaceConverter<M, R> {
+
+    public abstract boolean supportsModel(BrowsableDSpaceObject bdso);
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/BrowsableDSpaceObjectConverter.java]: Please add Javadocs for this abstract method

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkflowItemConverter.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        return witem;
+    }
+
+    @Override
+    public XmlWorkflowItem toModel(WorkflowItemRest obj) {
+        return null;
+    }
+
+
+    public void addError(List<ErrorRest> errors, ErrorRest toAdd) {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkflowItemConverter.java]: Please add Javadocs

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.dspace.app.rest.RestResourceController;
+
+/**
+ * The WorkflowItem REST Resource
+ *
+ * @author Andrea Bollini (andrea.bollini at 4science.it)
+ */
+public class ClaimedTaskRest extends BaseObjectRest<Integer> {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java]: Please improve Javadocs. I don't think this is a "WorkflowItem"

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    public String getType() {
+        return NAME;
+    }
+
+    @Override
+    public Class getController() {
+        return RestResourceController.class;
+    }
+
+    public String getStep() {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java]: Please add Javadocs to all, new public methods (which are not override)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/PoolTaskRest.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    public String getType() {
+        return NAME;
+    }
+
+    @Override
+    public Class getController() {
+        return RestResourceController.class;
+    }
+
+    public String getStep() {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/PoolTaskRest.java]: Please add Javadocs to all, new public methods (which are not override)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/WorkflowItemRest.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    public String getCategory() {
+        return CATEGORY;
+    }
+
+    @Override
+    public String getType() {
+        return NAME;
+    }
+
+    public Date getLastModified() {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/WorkflowItemRest.java]: Please add Javadocs to all, new public methods (which are not override)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java [hunk 18]
```diff
@@ -204,6 +204,18 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
         throw new RepositoryMethodNotImplementedException(apiCategory, model);
     }
 
+    public T action(HttpServletRequest request, ID id) throws SQLException, IOException, AuthorizeException {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: Please add Javadocs to all new public methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/submit/SubmissionService.java [hunk 19]
```diff
@@ -144,4 +163,51 @@ public UploadBitstreamRest buildUploadBitstream(ConfigurationService configurati
         return data;
     }
 
+    public XmlWorkflowItem createWorkflowItem(Context context, Request currentRequest)
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/submit/SubmissionService.java]: Please add Javadocs to all new public methods

File: dspace-api/src/main/java/org/dspace/browse/BrowsableDSpaceObject.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+ */
+package org.dspace.browse;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.dspace.core.Constants;
+
+public interface BrowsableDSpaceObject<PK extends Serializable> {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/browse/BrowsableDSpaceObject.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-2cd605c77b87a4381d059134ed6226b9R17

File: dspace-api/src/main/java/org/dspace/content/Bitstream.java [hunk 21]
```diff
@@ -426,4 +435,17 @@ public void setAcceptanceDate(Context context, DCDate acceptanceDate) throws SQL
             .setMetadataSingleValue(context, this, "dcterms", "accessRights", null, null, acceptanceDate.toString());
     }
 
+    public BrowsableDSpaceObject getParentObject() {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/Bitstream.java]: removed it was not needed https://github.com/DSpace/DSpace/pull/2312/files#diff-a5a6d29a63c46bed3ea222cf83a5b739

File: dspace-api/src/main/java/org/dspace/content/Bitstream.java [hunk 22]
```diff
@@ -426,4 +435,17 @@ public void setAcceptanceDate(Context context, DCDate acceptanceDate) throws SQL
...
+        Context context = new Context();
+        try {
+            return (BrowsableDSpaceObject) (getBitstreamService().getParentObject(context, this));
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+        }
+        return null;
+    }
+
+    public String getMetadata(String field) {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/Bitstream.java]: removed, unnecessary
https://github.com/DSpace/DSpace/pull/2312/files#diff-a5a6d29a63c46bed3ea222cf83a5b739

File: dspace-api/src/main/java/org/dspace/content/DSpaceObject.java [hunk 23]
```diff
@@ -199,4 +201,7 @@ protected void setModified() {
         this.modified = true;
     }
 
+    public String getUniqueIndexID() {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/DSpaceObject.java]: removed, the default implementation is enough (javadoc on the interface)

File: dspace-api/src/main/java/org/dspace/content/Item.java [hunk 24]
```diff
@@ -366,4 +376,9 @@ public ItemService getItemService() {
         }
         return itemService;
     }
+
+    public String getTypeText() {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/Item.java]: it is an override (now annotated). Javadoc on the interface

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/PoolTaskRest.java (hunk 16 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/model/PoolTaskRest.java]: it is just a DTO mapping the data model class, add link to the corresponding data model method/attribute https://github.com/DSpace/DSpace/pull/2312/files#diff-8afc90b30c652b5b004a16ef220c8124R46

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java (hunk 15 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java]: it is just a DTO mapping the data model class, add link to the corresponding data model method/attribute https://github.com/DSpace/DSpace/pull/2312/files#diff-8afc90b30c652b5b004a16ef220c8124R46

File: dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java (hunk 4 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-19084da5a2eab9b54275fe9030809c34R263

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java (hunk 5 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/service/ItemService.java]: removed unnecessary methods and add javadoc to the residual https://github.com/DSpace/DSpace/pull/2312/files#diff-ce065cd345298b0ebf3a369801eb7399R654

File: dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItem.java (hunk 6 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItem.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-537f4a26a5f610d5b958104155e9d36aR207

File: dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java (hunk 7 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java]: removed as related to the messaging feature to be discussed in a separate PR

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/ClaimedTask.java (hunk 8 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/ClaimedTask.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-8a2f0d69cfbc625b36f5d0da020e399bR129

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/PoolTask.java (hunk 9 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/PoolTask.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-52c496a60902e078268287bedbe773ffR144

File: dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java (hunk 11 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/storedcomponents/XmlWorkflowItem.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-1a554ce535923e7c211241af6f7385c8R158

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/BrowsableDSpaceObjectConverter.java (hunk 12 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/BrowsableDSpaceObjectConverter.java]: done https://github.com/DSpace/DSpace/pull/2312/files#diff-85f0e3672211829b8fceb596884bcdebR25

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkflowItemConverter.java (hunk 13 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkflowItemConverter.java]: it is (now) a private method https://github.com/DSpace/DSpace/pull/2312/files#diff-6e307cdd13b1a67847458f429e6bb747R139

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java (hunk 14 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/model/ClaimedTaskRest.java]: fixed https://github.com/DSpace/DSpace/pull/2312/files#diff-8afc90b30c652b5b004a16ef220c8124R15

**Comment by abollini** [comment]
discussion moved on the official PR https://github.com/DSpace/DSpace/pull/2312

**Pull request closed without merging**